src/V3MergeCond.cpp

```diff
@@ -51,71 +51,347 @@
+                        AstUser3Allocator<AstNodeStmt, StmtProperties>& stmtProperties) {
+        CodeMotionAnalysisVisitor{nodep, stmtProperties};
+    }
+};
+
+class CodeMotionOptimizeVisitor final : public AstNVisitor {
+    // NODE STATE
+    // AstNodeStmt::user3   -> StmtProperties (accessed via m_stmtProperties, maneged
+    //                         externally, see MergeCondVisitor::process)
+    // AstNodeStmt::user4   -> Flag: Already processed this node
+
+    AstUser4InUse m_user4InUse;
+
+    const AstUser3Allocator<AstNodeStmt, StmtProperties>& m_stmtProperties;
+
+    // MEMBERS
+
+    // Predicate to check if two sets are disjoint
+    static bool areDisjoint(const std::set<const AstVar*>& a, const std::set<const AstVar*>& b) {
+        if (a.empty() || b.empty()) return true;
+        const auto endA = a.end();
+        const auto endB = b.end();
+        auto itA = a.begin();
+        auto itB = b.begin();
+        for (;;) {
+            if (*itA == *itB) return false;
+            if (std::less<const AstVar*>{}(*itA, *itB)) {
+                itA = std::lower_bound(++itA, endA, *itB);
+                if (itA == endA) return true;
+            } else {
+                itB = std::lower_bound(++itB, endB, *itA);
+                if (itB == endB) return true;
+            }
+        }
+    }
+
+    // Predicate that checks if the order of two statements can be swapped
+    bool areSwappable(const AstNodeStmt* ap, const AstNodeStmt* bp) const {
+        StmtProperties& aProps = m_stmtProperties(ap);
+        StmtProperties& bProps = m_stmtProperties(bp);
+        // Don't move accross fences
+        if (aProps.m_isFence) return false;
+        if (bProps.m_isFence) return false;
+        // If either statement writes a variable that the other reads, they are not swappable
+        if (!areDisjoint(aProps.m_rdVars, bProps.m_wrVars)) return false;
+        if (!areDisjoint(bProps.m_rdVars, aProps.m_wrVars)) return false;
+        // If they both write to the same variable, they are not swappable
+        if (!areDisjoint(aProps.m_wrVars, bProps.m_wrVars)) return false;
+        // Otherwise good to go
+        return true;
+    }
+
+    // VISITORS
+    void visit(AstNode* nodep) override {}  // Ignore all non-statements
+
+    void visit(AstNodeStmt* nodep) override {
+        // Process only on first encounter
+        if (nodep->user4SetOnce()) return;
+        // First order children
+        iterateChildren(nodep);
+        // Grab hold of previous node with same condition
+        AstNodeStmt* prevp = m_stmtProperties(nodep).m_prevWithSameCondp;
+        // If no previous node with same condition, we are done
+        if (!prevp) return;
+#ifdef VL_DEBUG
+        {  // Sanity check, only in debug build, otherwise expensive
+            const AstNode* currp = prevp;
+            while (currp && currp != nodep) currp = currp->nextp();
+            UASSERT_OBJ(currp, nodep, "Predecessor not in same list");
```

Comment:
```suggestion
            UASSERT_OBJ(currp, nodep, "Predecessor not in same list as " << currp);
```

(If fires want to see both)